### PR TITLE
[CARBONDATA-792] Range Filter Optimization

### PR DESCRIPTION
**Pull request by sounakr**
Problem : LessThan and GreaterThan Expression are merged into a single RANGE expression.

Analysis : If LessThan and GreaterThan Expression on a same column given as filter predicates, then each expression is evaluated independently and finally AND operator is applied on individual results to get the final output. 

Fix : If the LessThan and GreaterThan Expression values can form a Range then it is Converted into a RANGE expression and results are evaluated through single expression. I.e. if the value lies within Range then choosen. This fix will optimize Block Prunning and Reduce Scanning time in case queries have range expressions in filter.

### CODE REVIEW

**Comment by CarbonDataQA** [comment]
Build Success with Spark 1.6.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/1222/

File: core/src/main/java/org/apache/carbondata/core/scan/filter/executer/RangeValueFilterExecuterImpl.java [hunk 1]
```diff
@@ -0,0 +1,557 @@
...
+    int startMin = 0;
+    int endMax = 0;
+    byte[][] filterValues = this.filterRangesValues;
+
+    // For Range expression we expect two values. The First is the Min Value and Second is the
+    // Max value.
+    // Get the Min Value
+    start = CarbonUtil
+        .getFirstIndexUsingBinarySearch(dimensionColumnDataChunk, startIndex, numerOfRows - 1,
+            filterValues[0], false);
```
Comment (gvramana) [inline core/src/main/java/org/apache/carbondata/core/scan/filter/executer/RangeValueFilterExecuterImpl.java]: Change the flag in binary search based on greaterthan case

File: core/src/main/java/org/apache/carbondata/core/scan/filter/executer/RangeValueFilterExecuterImpl.java [hunk 2]
```diff
@@ -0,0 +1,557 @@
...
+        start = start - 1;
+      }
+    }
+    endMax = start;
+
+    for (int j = startMin; j <= endMax; j++) {
+      bitSet.set(dimensionColumnDataChunk.getInvertedIndex(j));
+    }
+
+    if (dimensionColumnDataChunk.isNoDicitionaryColumn()) {
```
Comment (gvramana) [inline core/src/main/java/org/apache/carbondata/core/scan/filter/executer/RangeValueFilterExecuterImpl.java]: Binary search cannot be done on "@NU#LL$!", so need to check and compare for null on matching rows.

File: core/src/main/java/org/apache/carbondata/core/scan/expression/BinaryExpression.java [hunk 3]
```diff
@@ -38,4 +38,17 @@ public Expression getRight() {
     return right;
   }
 
+  @Override public void setChildren(Expression oldExpr, Expression newExpr) {
```
Comment (sounakr) [inline core/src/main/java/org/apache/carbondata/core/scan/expression/BinaryExpression.java]: Done

**Comment by CarbonDataQA** [comment]
Build Success with Spark 1.6.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/1242/

File: core/src/main/java/org/apache/carbondata/core/scan/filter/executer/RangeValueFilterExecuterImpl.java [hunk 4]
```diff
@@ -0,0 +1,557 @@
...
+    boolean greaterThan = isGreaterThan();
+    boolean greaterThanEqualTo = isGreaterThanEqualTo();
+    boolean lessThanEqualTo = isLessThanEqualTo();
+
+    int filterMinlessThanBlockMin =
+        ByteUtil.UnsafeComparer.INSTANCE.compareTo(blockMinValue, filterValues[0]);
+    int filterMaxLessThanBlockMin =
+        ByteUtil.UnsafeComparer.INSTANCE.compareTo(blockMinValue, filterValues[1]);
+
+    int filterMinGreaterThanBlockMax =
```
Comment (sounakr) [inline core/src/main/java/org/apache/carbondata/core/scan/filter/executer/RangeValueFilterExecuterImpl.java]: Done

File: core/src/main/java/org/apache/carbondata/core/scan/filter/executer/RangeValueFilterExecuterImpl.java [hunk 5]
```diff
@@ -0,0 +1,557 @@
...
+   */
+  public boolean isScanRequired(byte[] blockMinValue, byte[] blockMaxValue, byte[][] filterValues) {
+    boolean isScanRequired = true;
+
+    boolean lessThan = isLessThan();
+    boolean greaterThan = isGreaterThan();
+    boolean greaterThanEqualTo = isGreaterThanEqualTo();
+    boolean lessThanEqualTo = isLessThanEqualTo();
+
+    int filterMinlessThanBlockMin =
```
Comment (sounakr) [inline core/src/main/java/org/apache/carbondata/core/scan/filter/executer/RangeValueFilterExecuterImpl.java]: Done

File: core/src/main/java/org/apache/carbondata/core/scan/filter/executer/RangeValueFilterExecuterImpl.java [hunk 6]
```diff
@@ -0,0 +1,557 @@
...
+   * values and comparing them with filter min and max value.
+   * @param blockMinValue
+   * @param blockMaxValue
+   * @param filterValues
+   * @return
+   */
+  public boolean isScanRequired(byte[] blockMinValue, byte[] blockMaxValue, byte[][] filterValues) {
+    boolean isScanRequired = true;
+
+    boolean lessThan = isLessThan();
```
Comment (sounakr) [inline core/src/main/java/org/apache/carbondata/core/scan/filter/executer/RangeValueFilterExecuterImpl.java]: Done

File: core/src/main/java/org/apache/carbondata/core/scan/filter/executer/RangeValueFilterExecuterImpl.java [hunk 7]
```diff
@@ -0,0 +1,557 @@
...
+import org.apache.carbondata.core.scan.expression.exception.FilterUnsupportedException;
+import org.apache.carbondata.core.scan.filter.FilterUtil;
+import org.apache.carbondata.core.scan.filter.resolver.resolverinfo.DimColumnResolvedFilterInfo;
+import org.apache.carbondata.core.scan.filter.resolver.resolverinfo.MeasureColumnResolvedFilterInfo;
+import org.apache.carbondata.core.scan.processor.BlocksChunkHolder;
+import org.apache.carbondata.core.util.BitSetGroup;
+import org.apache.carbondata.core.util.ByteUtil;
+import org.apache.carbondata.core.util.CarbonUtil;
+
+public class RangeValueFilterExecuterImpl extends ValueBasedFilterExecuterImpl {
```
Comment (sounakr) [inline core/src/main/java/org/apache/carbondata/core/scan/filter/executer/RangeValueFilterExecuterImpl.java]: Done

File: core/src/main/java/org/apache/carbondata/core/scan/filter/executer/RangeValueFilterExecuterImpl.java [hunk 8]
```diff
@@ -0,0 +1,557 @@
...
+    BitSetGroup bitSetGroup = new BitSetGroup(rawColumnChunk.getPagesCount());
+    for (int i = 0; i < rawColumnChunk.getPagesCount(); i++) {
+      if (rawColumnChunk.getMaxValues() != null) {
+        if (isScanRequired(rawColumnChunk.getMinValues()[i], rawColumnChunk.getMaxValues()[i],
+            this.filterRangesValues)) {
+          if (isRangeFullyCovered(rawColumnChunk.getMinValues()[i],
+              rawColumnChunk.getMaxValues()[i], this.filterRangesValues)) {
+            // Set all the bits in this case as filter Min Max values cover the whole block.
+            BitSet bitSet = new BitSet(rawColumnChunk.getRowCount()[i]);
+            bitSet.flip(0, rawColumnChunk.getRowCount()[i]);
```
Comment (sounakr) [inline core/src/main/java/org/apache/carbondata/core/scan/filter/executer/RangeValueFilterExecuterImpl.java]: Here bitSet.flip parameters are startIndex and endIndex of the bits to flip. The new BitSet will give all bit initialized to 0. bitSet.flip will turn into 1.

File: core/src/main/java/org/apache/carbondata/core/scan/filter/executer/RangeValueFilterExecuterImpl.java [hunk 9]
```diff
@@ -0,0 +1,557 @@
...
+
+    int filterMinGreaterThanBlockMax =
+        ByteUtil.UnsafeComparer.INSTANCE.compareTo(filterValues[0], blockMaxValue);
+    int filterMaxGreaterThanBlockMax =
+        ByteUtil.UnsafeComparer.INSTANCE.compareTo(filterValues[1], blockMaxValue);
+
+    // if any filter value is in range than this block needs to be
+    // scanned less than equal to max range.
+
+    if (isDimensionPresentInCurrentBlock == true) {
```
Comment (sounakr) [inline core/src/main/java/org/apache/carbondata/core/scan/filter/executer/RangeValueFilterExecuterImpl.java]: Done

**Comment by CarbonDataQA** [comment]
Build Success with Spark 1.6.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/1257/

File: core/src/main/java/org/apache/carbondata/core/scan/expression/Expression.java [hunk 10]
```diff
@@ -42,6 +42,8 @@ public abstract ExpressionResult evaluate(RowIntf value)
     return children;
   }
 
+  public abstract void setChildren(Expression oldExpr, Expression newExpr);
```
Comment (sounakr) [inline core/src/main/java/org/apache/carbondata/core/scan/expression/Expression.java]: Done

File: core/src/main/java/org/apache/carbondata/core/scan/filter/executer/RangeValueFilterExecuterImpl.java (hunk 8 above)
Comment (sounakr) [inline core/src/main/java/org/apache/carbondata/core/scan/filter/executer/RangeValueFilterExecuterImpl.java]: It is set to 1.

File: core/src/main/java/org/apache/carbondata/core/scan/filter/executer/RangeValueFilterExecuterImpl.java [hunk 11]
```diff
@@ -0,0 +1,557 @@
...
+    boolean lessThan = isLessThan();
+    boolean greaterThan = isGreaterThan();
+
+    int filterMinlessThanBlockMin =
+        ByteUtil.UnsafeComparer.INSTANCE.compareTo(blockMinValue, filterValues[0]);
+
+    int filterMaxGreaterThanBlockMax =
+        ByteUtil.UnsafeComparer.INSTANCE.compareTo(filterValues[1], blockMaxValue);
+
+    if ((((greaterThan == true) && (filterMinlessThanBlockMin > 0)) || ((greaterThan == false) && (
```
Comment (sounakr) [inline core/src/main/java/org/apache/carbondata/core/scan/filter/executer/RangeValueFilterExecuterImpl.java]: Done

File: core/src/main/java/org/apache/carbondata/core/scan/filter/executer/RangeValueFilterExecuterImpl.java (hunk 1 above)
Comment (sounakr) [inline core/src/main/java/org/apache/carbondata/core/scan/filter/executer/RangeValueFilterExecuterImpl.java]: Done

**Comment by CarbonDataQA** [comment]
Build Success with Spark 1.6.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/1262/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 1.6.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/1279/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 1.6.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/1293/

File: core/src/main/java/org/apache/carbondata/core/scan/filter/intf/FilterOptimizerBasic.java [hunk 12]
```diff
@@ -0,0 +1,26 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.carbondata.core.scan.filter.intf;
+
+import org.apache.carbondata.core.scan.expression.Expression;
+
+public class FilterOptimizerBasic implements FilterOptimizer {
+  @Override public Expression optimizeFilter() {
```
Comment (sujith71955) [inline core/src/main/java/org/apache/carbondata/core/scan/filter/intf/FilterOptimizerBasic.java]: why this class required?

**Comment by gvramana** [comment]
LGTM

**Pull request closed without merging**